content/posts/building_jnigen.md

```diff
@@ -0,0 +1,262 @@
+
+## Why?
+
+The current way of accessing platform APIs on Flutter is Method Channels. Channels neatly avoid the intricacies of native language interop, by using an RPC like mechanism instead. However, there are a few drawbacks - all calls to method channels have to be asynchronous, and sharing memory is not possible. Perhaps more important, currently the use of method channels requires writing quite a bit of boilerplate. [^pigeon]
+
+Therefore, it's desirable to have an automatic bindings generator for Java libraries, for both ergonomics reasons. Using FFI instead of method channels enables better performance and eliminates the need for asynchrony as well.
+
+Dart already has an excellent bindings generator for C, called `ffigen`. It was written 2 years ago by Prerak Mann, in 2020 GSoC. It has been later extended to support Objective-C as well. It uses `libclang` to parse the C headers and generate wrapper code for calling them. The goal of `jnigen` was to provide similar facility for Java libraries.
+
+We had the prior art of `ffigen` which simplified several decisions. However there were challenges unique to Java.
+
+- How to parse the Java libraries? In C/C++, libclang is ubiquitous for this purpose.
+- In case of C, the `ffi` library is already provided by the dart. For Java, we had to develop a support library.
+- C dependency graphs are fairly sparse, and you can provide some header file paths to the tool. In comparison Java build systems and dependency configurations tend to be complex systems in their own right.
+
+As a rule of thumb, everything about Java interop (runtime, parsing, build) is less known, or less obvious. As a student, GSoC was my first attempt to write anything with real world scope. Therefore, I learned several practices and principles.
+
+The aim of this post is to give a glance of various architectural and design decisions, as well as the lessons I learned from this project.
+
+__TODO: A High level Architecture Diagram:__
+
+## JNI runtime support
+
+JNI is an old native interface [^jni]. Therefore, lot of quirks have to be considered when our goal is to abstract the JNI to a high level language like Dart.
+
+### Differences between platforms
+On Android, Flutter application runs embedded in Android JVM. Therefore the VM already exists. We just initialize the JNI using a plugin, and also obtain a reference to application context.
+
+On standalone targets (Flutter Desktop and standalone Dart), there's no JVM. On initialization, a JVM has to be spawned, using the JRE available on the machine. If you intend to use some external libraries, (like the PDFBox library we use in one of the example), the JAR files must be provided as classpath.
+
+Besides, we have a support library written in C. It's packaged automatically with Flutter apps since it's a native plugin. However, this library's path must be provided with standalone.
```

Comment:
Also, JNIgen supports ...

(I usually try to avoid "we" or "you", unless I want to draw attention to the people instead of the technical details.)